File: android/app/src/main/java/com/org/playboard/ui/matches/MatchesViewModel.kt
package com.org.playboard.ui.matches

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.org.playboard.data.auth.AuthRepository
import com.org.playboard.data.group.GroupRepository
import com.org.playboard.data.match.MatchRepository
import com.org.playboard.data.model.Group
import com.org.playboard.data.model.SessionState
import dagger.hilt.android.lifecycle.HiltViewModel
import java.time.LocalDate
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.distinctUntilChangedBy
import kotlinx.coroutines.flow.drop
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Matches tab (docs/requirements/03-matches.md): the active group's match log,
 * newest first. Cards expand in place to fetch full detail (breakdown + history);
 * matches can be deleted (with confirmation) by the recorder or a moderator.
 */
@HiltViewModel
class MatchesViewModel @Inject constructor(
    authRepository: AuthRepository,
    private val groupRepository: GroupRepository,
    private val matchRepository: MatchRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow(MatchesUiState())
    val uiState: StateFlow<MatchesUiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            authRepository.sessionState.collect { session ->
                _uiState.update { it.copy(currentUserId = (session as? SessionState.SignedIn)?.user?.id) }
            }
        }
        viewModelScope.launch {
            groupRepository.selectedGroup.distinctUntilChangedBy { it?.id }.collect { group ->
                if (group == null) {
                    _uiState.update {
                        it.copy(isLoading = false, noGroup = true, groupId = null, groupName = null, matches = emptyList())
                    }
                } else {
                    // New group: forget the previous group's per-day expand overrides so the
                    // list re-defaults to "newest day expanded, older days collapsed".
                    _uiState.update { it.copy(expandedDates = emptyMap()) }
                    loadMatches(group, showLoading = true)
                }
            }
        }
        // A match recorded/edited on the Add tab (or deleted) bumps the revision —
        // refresh silently so the list stays current without a spinner flash. An
        // edit also changes the open card's breakdown, so re-fetch that too.
        viewModelScope.launch {
            groupRepository.dataRevision.drop(1).collect {
                val group = groupRepository.selectedGroup.first() ?: return@collect
                loadMatches(group, showLoading = false)
                val expandedId = _uiState.value.expandedId
                if (expandedId != null) fetchDetail(group.id, expandedId)
            }
        }
    }

    fun retry() {
        viewModelScope.launch {
            val group = groupRepository.selectedGroup.first() ?: return@launch
            loadMatches(group, showLoading = true)
        }
    }

    /**
     * Pull-to-refresh: silently re-sync the group list (member counts) and this
     * group's match log, driving the pull indicator via [MatchesUiState.isRefreshing]
     * rather than the full-screen spinner. Also refreshes the open card's detail.
     */
    fun onPullRefresh() {
        viewModelScope.launch {
            _uiState.update { it.copy(isRefreshing = true) }
            groupRepository.refreshGroups(showLoading = false)
            groupRepository.selectedGroup.first()?.let { group ->
                loadMatches(group, showLoading = false)
                _uiState.value.expandedId?.let { fetchDetail(group.id, it) }
            }
            _uiState.update { it.copy(isRefreshing = false) }
        }
    }

    /** Expands a card (fetching its detail) or collapses it if already open. */
    fun onMatchClicked(matchId: String) {
        val state = _uiState.value
        if (state.expandedId == matchId) {
            _uiState.update { it.copy(expandedId = null, detail = null, isDetailLoading = false, detailFailed = false) }
            return
        }
        _uiState.update { it.copy(expandedId = matchId, detail = null, isDetailLoading = true, detailFailed = false) }
        val groupId = state.groupId ?: return
        viewModelScope.launch { fetchDetail(groupId, matchId) }
    }

    /** Loads a match's full detail into the expanded card (no-op if it's since collapsed). */
    private suspend fun fetchDetail(groupId: String, matchId: String) {
        matchRepository.getMatchDetail(groupId, matchId)
            .onSuccess { detail ->
                _uiState.update { if (it.expandedId == matchId) it.copy(isDetailLoading = false, detail = detail) else it }
            }
            .onFailure {
                _uiState.update { if (it.expandedId == matchId) it.copy(isDetailLoading = false, detailFailed = true) else it }
            }
    }

    fun onDeleteClicked(matchId: String) {
        _uiState.update { it.copy(deleteTargetId = matchId) }
    }

    fun onDeleteDismissed() {
        if (_uiState.value.isDeleting) return
        _uiState.update { it.copy(deleteTargetId = null) }
    }

    fun onDeleteConfirmed() {
        val state = _uiState.value
        val groupId = state.groupId ?: return
        val matchId = state.deleteTargetId ?: return
        _uiState.update { it.copy(isDeleting = true) }
        viewModelScope.launch {
            matchRepository.deleteMatch(groupId, matchId)
                .onSuccess {
                    // Optimistically drop it; the dataRevision bump also silently reloads.
                    _uiState.update {
                        it.copy(
                            isDeleting = false,
                            deleteTargetId = null,
                            matches = it.matches.filterNot { m -> m.id == matchId },
                            expandedId = if (it.expandedId == matchId) null else it.expandedId,
                            detail = if (it.expandedId == matchId) null else it.detail,
                        )
                    }
                }
                .onFailure { _uiState.update { it.copy(isDeleting = false, deleteTargetId = null) } }
        }
    }

    private suspend fun loadMatches(group: Group, showLoading: Boolean) {
        _uiState.update {
            it.copy(
                isLoading = showLoading,
                hasLoadFailed = false,
                noGroup = false,
                groupId = group.id,
                groupName = group.name,
                canModerate = group.myRole == "owner" || group.myRole == "admin",
            )
        }
        matchRepository.getMatches(group.id, mine = _uiState.value.showMineOnly)
            .onSuccess { page ->
                // First page always replaces the list and resets pagination to the top.
                _uiState.update {
                    it.copy(isLoading = false, matches = page.matches, nextCursor = page.nextCursor)
                }
            }
            .onFailure {
                // Keep a stale list on a silent refresh failure; only show the error
                // screen when there's nothing to show.
                _uiState.update { it.copy(isLoading = false, hasLoadFailed = it.matches.isEmpty()) }
            }
    }

    /**
     * Appends the next older page (triggered by the "Load older matches" footer). No-op
     * when a page is already loading or the log is fully loaded.
     */
    fun loadMore() {
        val state = _uiState.value
        val groupId = state.groupId ?: return
        val cursor = state.nextCursor ?: return
        if (state.isLoadingMore) return
        _uiState.update { it.copy(isLoadingMore = true) }
        viewModelScope.launch {
            matchRepository.getMatches(groupId, cursor = cursor, mine = state.showMineOnly)
                .onSuccess { page ->
                    _uiState.update {
                        val existingIds = it.matches.mapTo(HashSet()) { m -> m.id }
                        val appended = page.matches.filterNot { m -> m.id in existingIds }
                        it.copy(
                            isLoadingMore = false,
                            matches = it.matches + appended,
                            nextCursor = page.nextCursor,
                        )
                    }
                }
                // Keep the loaded list; the footer button stays for a manual retry.
                .onFailure { _uiState.update { it.copy(isLoadingMore = false) } }
        }
    }

    /** Toggles whether a day's section is expanded, recording an explicit user override. */
    fun onDateToggled(date: LocalDate) {
        _uiState.update { it.copy(expandedDates = it.expandedDates + (date to !it.isDateExpanded(date))) }
    }

    /**
     * Toggles the "My matches" filter and reloads the list from the top with the new scope.
     * Clears any open card since it may not be in the filtered result.
     */
    fun onToggleMineOnly() {
        _uiState.update {
            it.copy(
                showMineOnly = !it.showMineOnly,
                expandedId = null,
                detail = null,
                isDetailLoading = false,
                detailFailed = false,
            )
        }
        viewModelScope.launch {
            val group = groupRepository.selectedGroup.first() ?: return@launch
            loadMatches(group, showLoading = true)
        }
    }
}

File: ios/Playboard/Features/Matches/MatchesViewModel.swift
import Combine
import Foundation

/// Owns group-scoped pagination, filtering, refresh, and one expanded match at a time.
@MainActor
final class MatchesViewModel: ObservableObject {
    @Published private(set) var state = MatchesUiState()
    private let repository: any MatchRepository
    private var loadTask: Task<Void, Never>?
    private var detailTask: Task<Void, Never>?

    init(repository: any MatchRepository) { self.repository = repository }

    func select(group: PlayGroup?) {
        if state.groupID == group?.id {
            state.groupMatchCount = group?.matchCount ?? 0
            return
        }
        loadTask?.cancel()
        detailTask?.cancel()
        state = MatchesUiState(groupID: group?.id, groupMatchCount: group?.matchCount ?? 0)
        guard group != nil else { return }
        startLoad(showLoading: true)
    }

    func retry() { startLoad(showLoading: true) }

    func refresh() async {
        loadTask?.cancel()
        state.isRefreshing = true
        await load(showLoading: false)
        state.isRefreshing = false
    }

    func toggleMineOnly() {
        state.mineOnly.toggle()
        state.matches = []
        state.expandedMatchID = nil
        state.detail = nil
        startLoad(showLoading: true)
    }

    func loadMore() {
        guard let groupID = state.groupID, let cursor = state.nextCursor, !state.isLoadingMore else { return }
        let mineOnly = state.mineOnly
        state.isLoadingMore = true
        Task { [weak self] in
            guard let self else { return }
            do {
                let page = try await repository.matches(groupID: groupID, cursor: cursor, mine: mineOnly)
                try Task.checkCancellation()
                guard state.groupID == groupID, state.mineOnly == mineOnly else { return }
                let known = Set(state.matches.map(\.id))
                state.matches.append(contentsOf: page.matches.filter { !known.contains($0.id) })
                state.nextCursor = page.nextCursor
                state.isLoadingMore = false
            } catch is CancellationError {
                return
            } catch {
                state.isLoadingMore = false
                state.staleMessage = message(for: error)
            }
        }
    }

    func toggleDetail(matchID: String) {
        detailTask?.cancel()
        if state.expandedMatchID == matchID {
            state.expandedMatchID = nil
            state.detail = nil
            state.detailErrorMessage = nil
            return
        }
        guard let groupID = state.groupID else { return }
        state.expandedMatchID = matchID
        state.detail = nil
        state.isDetailLoading = true
        state.detailErrorMessage = nil
        detailTask = Task { [weak self] in
            guard let self else { return }
            do {
                let detail = try await repository.detail(groupID: groupID, matchID: matchID)
                try Task.checkCancellation()
                guard state.groupID == groupID, state.expandedMatchID == matchID else { return }
                state.detail = detail
                state.isDetailLoading = false
            } catch is CancellationError {
                return
            } catch {
                guard state.expandedMatchID == matchID else { return }
                state.isDetailLoading = false
                state.detailErrorMessage = message(for: error)
            }
        }
    }

    func sections(calendar: Calendar = .current) -> [MatchDaySection] {
        Dictionary(grouping: state.matches) { calendar.startOfDay(for: $0.playedAt) }
            .map { MatchDaySection(day: $0.key, matches: $0.value.sorted { $0.playedAt > $1.playedAt }) }
            .sorted { $0.day > $1.day }
    }

    private func startLoad(showLoading: Bool) {
        loadTask?.cancel()
        loadTask = Task { [weak self] in await self?.load(showLoading: showLoading) }
    }

    private func load(showLoading: Bool) async {
        guard let groupID = state.groupID else { return }
        let mineOnly = state.mineOnly
        if showLoading { state.isLoading = true }
        state.errorMessage = nil
        state.staleMessage = nil
        do {
            let page = try await repository.matches(groupID: groupID, cursor: nil, mine: mineOnly)
            try Task.checkCancellation()
            guard state.groupID == groupID, state.mineOnly == mineOnly else { return }
            state.matches = page.matches
            state.nextCursor = page.nextCursor
            state.isLoading = false
        } catch is CancellationError {
            return
        } catch {
            state.isLoading = false
            if state.matches.isEmpty { state.errorMessage = message(for: error) }
            else { state.staleMessage = "Showing saved matches. \(message(for: error))" }
        }
    }

    private func message(for error: Error) -> String {
        (error as? MatchRepositoryError)?.message ?? MatchRepositoryError.invalidResponse.message
    }
}
